ENH: ndimage: add array API standard support

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

<s>This is on top of https://github.com/scipy/scipy/pull/21091, so only the last commit (de94be0) is relevant.</s>

#### What does this implement/fix?
<!--Please explain your changes.-->

Allow `ndimage` functions consume and return arrays from API compliant namespaces.

- on cupy, delegate to `cupyx.scipy.ndimage` namesakes. See https://github.com/scipy/scipy/pull/21091 for the initial discussion.
- on other backends, cast inputs to numpy, do computations and cast the output back.

#### Additional information
<!--Any additional information you think is important.-->

Since `ndimage` is mainly compiled code with only a very small amount of python (mostly prepare/permute the inputs), we keep the original ndimage code almost intact.  Wrapping/unwrapping etc is done in a decorator.

ev-br commented on scipy/ndimage/_measurements.py:
```diff
@@ -618,77 +622,76 @@ def _stats(input, labels=None, index=None, centered=False):
-        means = sums / counts
-        centered_input = input - means[labels]
+        means = xp.astype(sums, xp.float64) / xp.astype(counts, xp.float64)   # XXX: f64 hardcoded
+        centered_input = xp.astype(input, means.dtype) - xp.take(means, labels)
```
This one is a fun one:  in numpy, `means[labels]`, is defined for `labels.ndim > 1` (in a weird way but hey); array-api-strict seems to only allow 1D labels.

rgommers commented on scipy/_lib/_array_api.py:
```diff
@@ -571,3 +587,62 @@ def xp_vector_norm(x: Array, /, *,
+
+
+def xp_ravel(a, xp):
+    return xp.reshape(a, (-1,))
```
I'd prefer to leave out `ravel` I think. The reshape one-liner is not only more or less the same in terms of readability, but it may also save a copy: `ravel` makes a copy if the array isn't C-ordered and contiguous, and in the usage here, that copy isn't needed.

rgommers commented on scipy/ndimage/_measurements.py:
```diff
@@ -618,82 +623,81 @@ def _stats(input, labels=None, index=None, centered=False):
-        bc = np.bincount(labels.ravel(),
-                              weights=(centered_input *
-                                       centered_input.conjugate()).ravel())
+        w = xp_ravel(centered_input * xp_conj(centered_input), xp)
```
Complex dtypes don't seem to be supported at all. Removing `.conjugate()` doesn't fail any tests. So doesn't seem needed?

rgommers commented on scipy/_lib/_array_api.py:
```diff
@@ -571,3 +587,62 @@ def xp_vector_norm(x: Array, /, *,
+
+
+# work around xp.real raising for floats etc
+def xp_conj(a, xp=None):
```
This one I can imagine could be an annoyance. However the one place this was used it didn't seem needed.

rgommers commented on scipy/ndimage/tests/test_fourier.py:
```diff
@@ -1,137 +1,174 @@
-        assert_almost_equal(ndimage.sum(a.real), 1.0, decimal=dec)
+        a = fft.ifft(a, n=shape[1], axis=1)
+        a = fft.ifft(a, n=shape[0], axis=0)
+        assert_almost_equal(ndimage.sum(xp_real(a)), xp.asarray(1.0), decimal=dec)
```
Just `xp.real` should work, since `a` is complex.

rgommers commented on scipy/ndimage/tests/test_fourier.py:
```diff
@@ -1,137 +1,174 @@
-        assert_almost_equal(ndimage.sum(a.real), 1.0, decimal=dec)
+        a = fft.ifft(a, n=shape[1], axis=1)
+        a = fft.ifft(a, n=shape[0], axis=0)
+        assert_almost_equal(ndimage.sum(xp_real(a)), xp.asarray(1.0), decimal=dec)
```
Same, `xp.real` should work.

rgommers commented on scipy/_lib/_array_api.py:
```diff
@@ -571,3 +587,62 @@ def xp_vector_norm(x: Array, /, *,
         return np.linalg.norm(x, ord=ord, axis=axis, keepdims=keepdims)
+
+
+# work around xp.real raising for floats etc
```
Doesn't seem needed as far as I can tell (see `xp_imag` below for a longer comment).

rgommers commented on scipy/ndimage/_measurements.py:
```diff
@@ -618,82 +623,81 @@ def _stats(input, labels=None, index=None, centered=False):
         return single_group(input[labels > 0])
 
-    if np.isscalar(index):
+    if xp_isscalar(index, xp):
```
Doing input validation on `index` and then here simply `index.ndim == 0` looks more clear.

rgommers commented on scipy/_lib/_array_api.py:
```diff
@@ -571,3 +587,62 @@ def xp_vector_norm(x: Array, /, *,
+
+
+def xp_isscalar(a, xp):
+    return xp.asarray(a).ndim == 0
```
This is incorrect, e.g.:
```python
>>> def xp_isscalar(a, xp=np):
...     return xp.asarray(a).ndim == 0
...
>>> np.isscalar(None)
False
>>> xp_isscalar(None)
True
```

`np.isscalar` is not a great function. Code that uses it now should usually use `a.ndim == 0`, which works for all array types and for numpy scalars. Please don't add it here.

rgommers commented on scipy/_lib/_array_api.py:
```diff
@@ -571,3 +587,62 @@ def xp_vector_norm(x: Array, /, *,
+        return xp.asarray(result)
+
+
+def xp_searchsorted(x1, x2, xp=None):
```
`array-api-strict` now has `searchsorted`, which I assume was the problem here. So probably no longer needed?

rgommers reviewed (commented):
Thanks @ev-br!

This is a huge PR that looks good at a high level. I only reviewed the array API usage and the new utility wrappers/functions added in `_lib/_array_api.py`. Some of those indeed seem useful and make things a lot easier (the testing functions in particular), but some seem unnecessary. It's better to use idiomatic functions where possible, rather than rewrite everything that we had as `np.xxx` as `xp_xxx` (e.g., `xp_imag`, `xp_isscalar`). So I'd like to see those removed. The resulting code changes should actually (a) be minor, and (b) make the code more readable.

ev-br commented:
> I only reviewed the array API usage and the new utility wrappers/functions added in _lib/_array_api.py

Thanks Ralf! Have to admit I don't like where this is going, too, and I'm going to rework it all to avoid `xp_*` additions at an expense of some more meta-stuff.

lucascolley commented on scipy/ndimage/tests/test_interpolation.py:
```diff
@@ -538,418 +654,459 @@ def test_map_coordinates_endianness_with_output_parameter(self):
+        ndimage.affine_transform(data, xp.asarray([[1]]), output=out)
         assert_array_almost_equal(out[:8], data)
 
         # mismatched output shape raises an error
         with pytest.raises(RuntimeError):
             ndimage.affine_transform(
                 data, [[1]], output=out, output_shape=(12,))
 
-    def test_affine_transform_with_string_output(self):
-        data = np.array([1])
-        out = ndimage.affine_transform(data, [[1]], output='f')
-        assert_(out.dtype is np.dtype('f'))
-        assert_array_almost_equal(out, [1])
+    def test_affine_transform_with_string_output(self, xp):
+        if not is_numpy(xp):
+            pytest.xfail("string output is numpy-specific")
+
+        data = xp.asarray([1])
+        out = ndimage.affine_transform(data, xp.asarray([[1]]), output='f')
+        assert out.dtype is np.dtype('f')
+        assert_array_almost_equal(out, xp.asarray([1]))
 
     @pytest.mark.parametrize('shift',
                              [(1, 0), (0, 1), (-1, 1), (3, -5), (2, 7)])
     @pytest.mark.parametrize('order', range(0, 6))
-    def test_affine_transform_shift_via_grid_wrap(self, shift, order):
+    def test_affine_transform_shift_via_grid_wrap(self, shift, order, xp):
         # For mode 'grid-wrap', integer shifts should match np.roll
-        x = np.array([[0, 1],
+        x = xp.asarray([[0, 1],
                       [2, 3]])
-        affine = np.zeros((2, 3))
-        affine[:2, :2] = np.eye(2)
-        affine[:, 2] = shift
+        affine = xp.zeros((2, 3))
+        affine[:2, :2] = xp.eye(2)
+        affine[:, 2] = xp.asarray(shift)
+
+        expected = np.roll(np.asarray(x), shift, axis=(0, 1))
+        expected = xp.asarray(expected)
```
https://github.com/scipy/scipy/pull/21091#discussion_r1676804014 is being deferred to this PR and should be fixed here - no need to `xp.asarray` then  `np.asarray` `x` before it is used.

ev-br replied:
This has been fixed: https://github.com/scipy/scipy/pull/21150/files#diff-3a88694edba2c070f37a8ec5f066aa515ae86d39a2f971fe74fe97e4c1579d21L881

ev-br commented:
I'm probably missing something very simple. So the point of `array_api_strict` is to provide a lowest common denominator for all array API libraries, and if tests pass with `-b array_api_strict`, they should pass under all backends, correct?

However IRL, after an update which makes tests pass under `array_api_strict == 2`, other backends fail left right and center with missing array API attributes and methods:

```
$ python dev.py test -s ndimage -b cupy -j 12
...
FAILED scipy/ndimage/tests/test_measurements.py::test_sum10[cupy] - AttributeError: module 'cupy' has no attribute 'bool'
FAILED scipy/ndimage/tests/test_interpolation.py::TestNdimageInterpolation::test_affine_transform26[2-cupy] - AttributeError: module 'cupy' has no attribute 'concat'
...

Am I missing something simple?

```

lucascolley commented:
> However IRL, after an update which makes tests pass under array_api_strict == 2, other backends fail left right and center with missing array API attributes and methods:

`array_api_compatible` parametrises tests with the raw (i.e. un-array_api_compat-wrapped) namespaces. If you need to use functions which exist in the standard but not in the main namespaces of the libraries we test with, you can use the pattern.

```python
xp_test = array_namespace(x) # explain here why we need the wrapped namespace, e.g. # `cupy` needs `concat`
# use `xp_test` instead of `xp`
```

You should ensure that the arrays which are passed into the functions being tested are still created with the raw namespace `xp` rather than `xp_test`.

Sometimes it is better to avoid using this pattern by using NumPy functions and converting to `xp` at a later point.

ev-br commented on scipy/ndimage/tests/test_fourier.py:
```diff
@@ -1,137 +1,174 @@
+        a = fft.ifft(a, n=shape[1], axis=1)
+        a = fft.ifft(a, n=shape[0], axis=0)
+        assert_array_almost_equal(xp_real(a)[1:, 1:], expected[:-1, :-1], decimal=dec)
+        assert_array_almost_equal(xp_imag(a), xp.zeros(shape), decimal=dec)
```
`xp_imag` is gone in the updated version, together with other xp_ additions.

ev-br commented:
Here is an update to this PR:

- original ndimage functions are NumPy-specific; they convert their inputs to numpy as before and return numpy arrays;
- publicly exposed functions are decorated for the dispatch/delegation, and the decorator converts outputs from numpy to the target namespace;
- the only changes to the original ndimage code are a handful `np.asarray`, where previously missing;
- `xp_` additions are gone
- the whole of ndimage should work with all backends;
- remaining CI failures are mostly 50-odd instances of https://github.com/scipy/scipy/issues/21044
- the bulk of this PR is tweaks to `ndimage/testing` from https://github.com/scipy/scipy/pull/21091, the actual diff with respect to that PR is very small.

ev-br commented:
> the whole of ndimage should work with all backends;

Several caveats to mention upfront:
- several ndimage routines use in-place operations internally, which does not work with JAX. Changing these to be out-of-place is likely pessimizing performance for everyone else, so is not done here. 
- `output=dtype` only works on NumPy and CuPy--- the API is that `if isinstance(output, (np.dtype, np.generic))`, the output buffer is freshly allocated with this dtype. Making this work for non-numpy dtypes is not hard: the only thing we'd need is a generic way to map `xp`-namespace dtypes to numpy dtypes. Whether it's worth it, I'm not sure; have not done it for now. Maybe should see some usage first. OTOH, raising a useful error is the same amount of effort as adding the proper support.
- `output=xp_array` works on NumPy and CuPy, and not other backends. The failure mode is unpleasant: there is no extra allocation (good) but the `output` argument is not updated in-place (bad). Detecting this and raising an error is about the same amount of work as adding a proper support.

ev-br commented:
CI failures are https://github.com/scipy/scipy/issues/21044 again. Note that the last commit does what the error message suggests --- and it just transfers the breakage from `-b numpy` runs to "normal" runs where `SCIPY_ARRAY_API` variable is not defined.

lucascolley commented on scipy/_lib/_array_api.py:
```diff
@@ -381,6 +381,24 @@ def xp_assert_less(actual, desired, check_namespace=True, check_dtype=True,
                                         err_msg=err_msg, verbose=verbose)
 
 
+def assert_array_almost_equal(actual, desired, decimal=6, *args, **kwds):
+    """Backwards compatible replacement. In new code, use xp_assert_close instead.
+    """
+    rtol, atol = 0, 1.5*10**(-decimal)
+    return xp_assert_close(actual, desired,
+                           atol=atol, rtol=rtol, check_dtype=False, allow_0d=True,
+                           *args, **kwds)
+
+
+def assert_almost_equal(actual, desired, decimal=7, *args, **kwds):
+    """Backwards compatible replacement. In new code, use xp_assert_close instead.
+    """
+    rtol, atol = 0, 1.5*10**(-decimal)
+    return xp_assert_close(actual, desired,
+                            atol=atol, rtol=rtol, check_dtype=False, allow_0d=True,
+                            *args, **kwds)
```
let's make these typed since the rest of the file is. Also indentation on lines 398, 399

ev-br replied:
`xp_assert_*` functions are not: https://github.com/scipy/scipy/blob/main/scipy/_lib/_array_api.py#L312

lucascolley replied:
Apologies, you're right (https://github.com/scipy/scipy/pull/20621#issue-2273073907)

lucascolley commented on scipy/ndimage/tests/test_splines.py:
```diff
@@ -1,11 +1,18 @@
+skip_xp_backends = pytest.mark.skip_xp_backends
+pytestmark = [array_api_compatible, pytest.mark.usefixtures("skip_xp_backends"),
+              # XXX: only CuPy delegation is implemented
+              # skip_xp_backends("jax.numpy", "torch"),
```
what is the story with JAX and PyTorch now?

ev-br replied:
Yup, removed.

ev-br commented on scipy/_lib/_array_api.py:
```diff
@@ -381,6 +381,24 @@ def xp_assert_less(actual, desired, check_namespace=True, check_dtype=True,
+    """
+    rtol, atol = 0, 1.5*10**(-decimal)
+    return xp_assert_close(actual, desired,
+                            atol=atol, rtol=rtol, check_dtype=False, allow_0d=True,
```
```suggestion
                           atol=atol, rtol=rtol, check_dtype=False, allow_0d=True,
```

ev-br commented on scipy/_lib/_array_api.py:
```diff
@@ -381,6 +381,24 @@ def xp_assert_less(actual, desired, check_namespace=True, check_dtype=True,
+    rtol, atol = 0, 1.5*10**(-decimal)
+    return xp_assert_close(actual, desired,
+                            atol=atol, rtol=rtol, check_dtype=False, allow_0d=True,
+                            *args, **kwds)
```
```suggestion
                            args, **kwds)
```

ev-br commented on scipy/ndimage/tests/test_splines.py:
```diff
@@ -1,11 +1,18 @@
+skip_xp_backends = pytest.mark.skip_xp_backends
+pytestmark = [array_api_compatible, pytest.mark.usefixtures("skip_xp_backends"),
+              # XXX: only CuPy delegation is implemented
+              # skip_xp_backends("jax.numpy", "torch"),
```
```suggestion
```

ev-br commented on scipy/ndimage/tests/test_splines.py:
```diff
@@ -1,11 +1,18 @@
+from scipy.conftest import array_api_compatible
+skip_xp_backends = pytest.mark.skip_xp_backends
+pytestmark = [array_api_compatible, pytest.mark.usefixtures("skip_xp_backends"),
+              # XXX: only CuPy delegation is implemented
```
```suggestion

```

lucascolley commented:
`boost_math` submodule update has crept in

ev-br commented:
> boost_math submodule update has crept in

Thanks, removed.

ev-br commented:
https://github.com/scipy/scipy/pull/21150/commits/a8072a02392006446c2faf1ceef435d5c17c75ea does a minimal fix to get the CI to be green. A structural solution from https://github.com/scipy/scipy/issues/21044 would be great but does not need to block this PR. We can always tighten ndimage test assertions, when their improvements land.

lucascolley commented on scipy/ndimage/_measurements.py:
```diff
@@ -1535,10 +1541,11 @@ def center_of_mass(input, labels=None, index=None):
     >>> ndimage.center_of_mass(d)
     (inf,)
     """
-    normalizer = sum(input, labels, index)
+    input = np.asarray(input)
+    normalizer = sum_labels(input, labels, index)
     grids = np.ogrid[[slice(0, i) for i in input.shape]]
 
-    results = [sum(input * grids[dir].astype(float), labels, index) / normalizer
+    results = [sum_labels(input * grids[dir].astype(float), labels, index) / normalizer
```
could you explain the change here?

ev-br replied:
This is pure cosmetics: `ndimage.sum` is an backwards compat alias for `sum_labels`, so use a specific name and avoid shadowing the builtin.

lucascolley commented on scipy/ndimage/_measurements.py:
```diff
@@ -507,6 +508,8 @@ def labeled_comprehension(input, labels, index, func, out_dtype, default,
         else:
             return func(input.ravel(), positions.ravel())
 
+    labels = np.asarray(labels)
```
was the need for these extra `np.asarray`'s caught by the new testing?

ev-br replied:
Yes. Otherwise e.g. torch arrays end up in numpy-specific ndimage code which does not end well.

lucascolley commented on scipy/ndimage/_support_alternative_backends.py:
```diff
@@ -0,0 +1,72 @@
+        def wrapper(*args, **kwds):
+            xp = dispatcher(*args, **kwds)
+
+            # try delegating to a cupyx/jax namesake
```
```suggestion
            # try delegating to a cupyx/jax namespace
```

ev-br replied:
Why? It does not delegate to a _namespace_, it delegates to an identically-named function from a namespace.

lucascolley replied:
okay :)

lucascolley commented on scipy/ndimage/tests/test_fourier.py:
```diff
@@ -1,137 +1,181 @@
+        a = fft.ifft(a, n=shape[1], axis=1)
+        a = fft.irfft(a, n=shape[0], axis=0)
         assert_array_almost_equal(a[1:, 1:], expected[:-1, :-1], decimal=dec)
-        assert_array_almost_equal(a.imag, np.zeros(shape), decimal=dec)
```
why is this removed?

ev-br replied:
https://github.com/scipy/scipy/pull/21150#discussion_r1675793114

lucascolley commented on scipy/ndimage/tests/test_fourier.py:
```diff
@@ -1,137 +1,181 @@
 from scipy import ndimage
 
+from scipy.conftest import array_api_compatible
+skip_xp_backends = pytest.mark.skip_xp_backends
+pytestmark = [array_api_compatible, pytest.mark.usefixtures("skip_xp_backends")]
```
looks like the skip infrastructure is unused here? Not sure why the linter hasn't picked this up.

ev-br replied:
Keep linter improvements for a follow-up then?

lucascolley reviewed (commented):
reviewed everything apart from the 4 big test files

lucascolley commented:
One more thing - could you add some array-like tests along the lines of gh-21209 ?

ev-br commented:
> One more thing - could you add some array-like tests along the lines of gh-21209 ?

To clarify, you're asking to add ~100 tests to an already ~5 kLOC diff?

lucascolley commented:
> > One more thing - could you add some array-like tests along the lines of https://github.com/scipy/scipy/pull/21209 ?
> 
> To clarify, you're asking to add ~100 tests to an already ~5 kLOC diff?

Perhaps better for a follow-up, then.

lucascolley commented on scipy/ndimage/tests/test_interpolation.py:
```diff
@@ -54,128 +65,152 @@ def shift(x):
+                                        output_shape=(4,)),
+            xp.asarray(expected_value))
 
+    @skip_xp_backends("torch", reasons=["ValueError: only one element tensors "
+                                         "can be converted to Python scalars"])
```
I should take a look at the traceback for this

lucascolley replied:
this diff fixes the torch errors:

```diff
diff --git a/scipy/ndimage/tests/test_interpolation.py b/scipy/ndimage/tests/test_interpolation.py
index 0074ad6dda..dc6ed0c7b1 100644
--- a/scipy/ndimage/tests/test_interpolation.py
+++ b/scipy/ndimage/tests/test_interpolation.py
@@ -75,8 +75,6 @@ class TestNdimageInterpolation:
                                         output_shape=(4,)),
             xp.asarray(expected_value))
 
-    @skip_xp_backends("torch", reasons=["ValueError: only one element tensors "
-                                         "can be converted to Python scalars"])
     @pytest.mark.parametrize('mode', ['mirror', 'reflect', 'grid-mirror',
                                       'grid-wrap', 'grid-constant',
                                       'nearest'])
@@ -92,14 +90,15 @@ class TestNdimageInterpolation:
         np_data = np.arange(-6, 7, dtype=np.float64)
         data = xp.asarray(np_data)
         x = xp.asarray(np.linspace(-8, 15, num=1000))
-        y = ndimage.map_coordinates(data, xp.asarray([x]), order=order, mode=mode)
+        newaxis = array_namespace(x).newaxis
+        y = ndimage.map_coordinates(data, x[newaxis, ...], order=order, mode=mode)
 
         # compute expected value using explicit padding via np.pad
         npad = 32
         pad_mode = ndimage_to_numpy_mode.get(mode)
         padded = xp.asarray(np.pad(np_data, npad, mode=pad_mode))
-        expected = ndimage.map_coordinates(padded, xp.asarray([npad + x]), order=order,
-                                           mode=mode)
+        coords = xp.asarray(npad + x)[newaxis, ...]
+        expected = ndimage.map_coordinates(padded, coords, order=order, mode=mode)
 
         atol = 1e-5 if mode == 'grid-constant' else 1e-12
         xp_assert_close(y, expected, rtol=1e-7, atol=atol)

```

lucascolley commented on scipy/ndimage/tests/test_interpolation.py:
```diff
@@ -403,51 +458,60 @@ def mapping(x, a, b):
             rtol=1e-7,
         )
 
     def test_geometric_transform_endianness_with_output_parameter(self):
+        # byteorder is only supported on NumPy
```
we should `np_only` skip here

lucascolley commented on scipy/ndimage/tests/test_interpolation.py:
```diff
@@ -538,518 +625,584 @@ def test_map_coordinates_endianness_with_output_parameter(self):
+        offset_original = -xp.ones((2, 1))
+
+        concat = array_namespace(tform_original, offset_original).concat
+        tform_h1 = concat((tform_original, offset_original), axis=1)  # hstack
+        tform_h2 = concat( (tform_h1, xp.asarray([[0.0, 0, 1]])), axis=0)  # vstack
```
can we use [`xp.stack`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.stack.html) for either of these?

ev-br replied:
Possibly. Is something wrong with what's here now though?

lucascolley replied:
this is okay 👍 let's fix the whitespace though
```suggestion
        tform_h1 = concat((tform_original, offset_original), axis=1)  # hstack
        tform_h2 = concat((tform_h1, xp.asarray([[0.0, 0, 1]])), axis=0)  # vstack
```

lucascolley commented on scipy/ndimage/tests/test_interpolation.py:
```diff
@@ -538,518 +625,584 @@ def test_map_coordinates_endianness_with_output_parameter(self):
-                      [2, 3]])
+        x = np.asarray([[0, 1],
+                        [2, 3]])
+        expected =  np.roll(x, shift, axis=(0,1))
```
```suggestion
        expected = np.roll(x, shift, axis=(0,1))
```

lucascolley commented on scipy/ndimage/tests/test_interpolation.py:
```diff
@@ -1061,267 +1214,289 @@ def test_shift_nearest_boundary(self, order, prefilter):
+        data = xp.asarray([[[0, 0, 0, 0, 0],
+                            [0, 1, 1, 0, 0],
+                            [0, 0, 0, 0, 0]]] * 2, dtype=xp.float64)
+        permute_dims = array_namespace(data).permute_dims
+        data = permute_dims(data, (2, 1, 0))  # == np.transpose
```
This pattern seems okay when there is only one function to be accessed 👍. `xp_test` is probably cleaner if multiple functions are needed, though.

lucascolley reviewed (commented):
this is looking really good now, not many comments from one of the big files

ev-br commented:
Thanks @lucascolley. Are you planning to review other test modules? I'd then wait for the full review and address all comments in one go.

lucascolley commented:
> Thanks @lucascolley. Are you planning to review other test modules? I'd then wait for the full review and address all comments in one go.

Yep, sounds good.

lucascolley commented on scipy/ndimage/tests/test_measurements.py:
```diff
@@ -101,236 +116,259 @@ def test_basic(self):
-    data = np.ones([5])
+def test_label_output_wrong_size(xp):
+    if is_jax(xp):
+        pytest.xfail("JAX does not raise")
+
+    data = xp.ones([5])
     for t in types:
-        output = np.zeros([10], t)
-        assert_raises((RuntimeError, ValueError),
+        dtype = getattr(xp, t)
+        output = xp.zeros([10], dtype=dtype)
+        # TypeError is from non-numpy arrays as output
+        assert_raises((ValueError, TypeError),
                       ndimage.label, data, output=output)
```
this is totally fine, but for the future, I would personally leave tests for specific errors as `np_only` unless they are mentioned under `Raises` in the docs - they can always be changed in the future, it just seems a bit unnecessary to test that at this stage imo

ev-br replied:
Agreed.

lucascolley commented on scipy/ndimage/tests/test_measurements.py:
```diff
@@ -340,682 +378,793 @@ def test_label_structuring_elements():
-    assert_(label.dtype in (np.int32, np.int64))
-    # Shouldn't raise an exception
-    ndimage.find_objects(label)
+# cupyx.scipy.ndimage does not have find_object
+@skip_xp_backends("cupy")
```
let's use `reasons=`

lucascolley commented on scipy/ndimage/tests/test_measurements.py:
```diff
@@ -1026,190 +1175,216 @@ def test_extrema02():
-    expected1 = [0, 1, 0, 1, 1]
-    expected2 = [0, 0, 0, 3, 0]
-    input = np.array([1, 1, 3, 4, 3, 5, 3, 3])
+@skip_xp_backends(np_only=True, reasons=['object arrays'])
+def test_histogram03(xp):
+    labels = xp.asarray([1, 0, 1, 1, 2, 2, 2, 2])
+    expected1 = xp.asarray([0, 1, 0, 1, 1])
+    expected2 = xp.asarray([0, 0, 0, 3, 0])
+    input = xp.asarray([1, 1, 3, 4, 3, 5, 3, 3])
+
     output = ndimage.histogram(input, 0, 4, 5, labels, (1, 2))
```
is this due to the tuple `(1, 2)`? If so, I think this should be resolved in the implementation - the docs say `index` can be a _sequence of ints_, so we should aim for that to work.

ev-br replied:
Possibly. Don't know TBH, am surprised by object arrays, too. Am not volunteering to fix the implementation though :-).

lucascolley replied:
Okay, my concern was a mishandling of the input tuple, but in fact the *result* is an object array of arrays. That is bizarre! Let's just leave this alone...

lucascolley replied:
Actually, no - since this is documented to accept a "sequence of ints" and return an "ndarray", we shoudn't be producing object arrays here which are incompatible with alternative backends. I'll mark this for follow-up, we might be in need of a deprecation cycle.

lucascolley commented on scipy/ndimage/tests/test_measurements.py:
```diff
@@ -1026,190 +1175,216 @@ def test_extrema02():
+    xp_assert_equal(max, xp.asarray([9, 5]), check_dtype=False)
 
 
+# XXX: cannot just @skip_xp_backend("cupy"): multiple skips do not add
```
I think this comment is outdated now

lucascolley commented on scipy/ndimage/tests/test_measurements.py:
```diff
@@ -1271,147 +1446,153 @@ def test_watershed_ift03(self):
+        assert_array_almost_equal(out, xp.asarray(expected))
 
-    def test_watershed_ift09(self):
+    @skip_xp_backends("cupy", "pytorch", reasons=["torch.uint16"])
```
this looks unintentional - the length of `reasons` should match the amount of `backends` (see https://github.com/scipy/scipy/blob/61f2a11cfdc07a7d167bd715aae6fe76d04c354e/scipy/conftest.py#L179-L205)

ev-br replied:
The whole class is skipped under CuPy, and there is a need to skip a specific test method --- et voila, the skip machinery is what it is.
I've duplicated the cupy reason both ways for now.

lucascolley reviewed (commented):
another file done

lucascolley commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -784,29 +966,36 @@ def test_filter_axes_kwargs(self, filter_func, size0, size, kwargs, axes):
-        axes = np.array(axes)
+        array = xp.arange(6 * 8 * 12, dtype=xp.float64)
+        array = xp.reshape(array, (6, 8, 12))
+        # axes = xp.asarray(axes)
```
If this works commented out, shall we just remove it?

lucascolley commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -1336,95 +1638,108 @@ def test_minmax_nonseparable_axes(self, filter_func, axes, kwargs):
+                            [5, 8, 3, 7, 1],
+                            [5, 6, 9, 3, 5]])
+
+        asarray = array_namespace(array).asarray
+        array_copy = asarray(array, copy=True)
```
we can use `_lib._array_api._asarray` here instead

ev-br replied:
Is the present version bad somehow. It works across the CI + locally on GPU, it looks obviously correct from a quick glance, what is wrong with it? 
A suggested alternative has three leading underscores, and I've no idea without testing what it's going to do under plain numpy.

lucascolley replied:
I don't think the present version is "bad", it's just "not quite as good" as using the helpers which are used elsewhere to abstract away the `array_namespace` call. Recalling the argument along the lines of 'some maintainers don't care about this stuff, so let's try to be minimally intrusive in the code'. This isn't a particularly good example since your pattern was short anyway, but I think it's generally good practice to use existing abstractions rather than reinvent the wheel.

> A suggested alternative has three leading underscores, and I've no idea without testing what it's going to do under plain numpy.

Well `_lib._array_api.array_namespace` has 2 leading underscores, so I don't really see the issue 🙃. If you would like to know what the function will do, take a look at [the docstring](https://github.com/scipy/scipy/blob/a58f4b61557b9ca649fd63d05338eee368efce41/scipy/_lib/_array_api.py#L132-L155)!

lucascolley replied:
All that to say that I don't think you need to change anything for me to merge this - just trying to encourage best practice.

lucascolley reviewed (commented):
full look-through done, fire away @ev-br !

ev-br commented on scipy/ndimage/tests/test_morphology.py:
```diff
@@ -885,104 +1017,136 @@ def test_binary_erosion26(self, dtype):
-                         [0, 0, 0, 1, 0, 0, 0],
-                         [0, 0, 0, 0, 0, 0, 0]], bool)
+        expected = xp.asarray(expected)
+        data = np.asarray([[0, 0, 0, 0, 0, 0, 0],
+                           [0, 0, 0, 1, 0, 0, 0],
+                           [0, 0, 1, 1, 1, 0, 0],
+                           [0, 1, 1, 1, 1, 1, 0],
+                           [0, 0, 1, 1, 1, 0, 0],
+                           [0, 0, 0, 1, 0, 0, 0],
+                           [0, 0, 0, 0, 0, 0, 0]], dtype=bool)
+        data = xp.asarray(data)
```
Here (and elsewhere) the pattern is not just to reduce the diff actually. This is deliberate to work around `xp.bool` being not usable:

```
>>> np.arange(3, dtype=np.bool)
<stdin>:1: FutureWarning: In the future `np.bool` will be defined as the corresponding NumPy scalar.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/br/miniforge3/envs/scipy-dev/lib/python3.12/site-packages/numpy/__init__.py", line 324, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'bool'.
`np.bool` was a deprecated alias for the builtin `bool`. To avoid this error in existing code, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations. Did you mean: 'bool_'?
>>> np.__version__
'1.26.4'
```

Thus convert to numpy first with `dtype=bool`, then `xp.asarray`.

lucascolley replied:
I did understand that - my point was that the indentation changed unnecessarily by going from `np.array` to `np.asarray`. Anyway, not an important critique, just a drive-by observation 😅

ev-br commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -385,326 +428,450 @@ def test_correlate17(self):
+        dtype_kernel = getattr(xp, dtype_kernel)
+        dtype_output = getattr(xp, dtype_output)
+
+        if is_cupy(xp) and mode == 'grid-constant':
+            pytest.xfail('https://github.com/cupy/cupy/issues/8404')
+        elif not is_numpy(xp):
+            pytest.xfail("output= arrays are numpy-specific")
```
Sure I'm all +1 if you know how to improve the infra; FWIW I don't see how , especially with extra conditions (`mode == 'grid-constant' here)

lucascolley replied:
> especially with extra conditions (`mode == 'grid-constant' here)

I don't think we need that level of detail - xfails for specific arguments would stay in the function. The part that should be abstracted out is the `not (is_numpy(xp) or is_cupy(xp))`.

ev-br commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -385,326 +428,450 @@ def test_correlate17(self):
+        o_sum = xp.sum(output, dtype=xp.float64)
+        i_sum = xp.sum(input, dtype=xp.float64)
+        assert_almost_equal(o_sum, i_sum, decimal=0)
+        assert sumsq(np.asarray(input), np.asarray(output)) > 1.0
```
Ah, indeed! However the test is xfailed on CuPy and has no chance to run under other GPU libs.

lucascolley replied:
Sure, but I'm not convinced it is a good idea to write a test which will fail as soon as the xfail reason is resolved (maybe it *won't* get resolved any time soon, but that is besides the point IMO). `sumsq` can easily be written to be array-agnostic:

```python
def sumsq(a, b, xp=None):
    xp = array_namespace(a, b) if xp is None else xp
    return xp.sqrt(xp.sum((a - b)**2))
```

Then call it with `xp=xp` to avoid extra `array_namespace` overhead.

We shouldn't need to be concerned by any slow down from `math.sqrt` to `xp.sqrt` as this is just testing code.

ev-br commented on scipy/ndimage/tests/test_measurements.py:
```diff
@@ -1,23 +1,28 @@
+skip_xp_backends = pytest.mark.skip_xp_backends
+pytestmark = [array_api_compatible, pytest.mark.usefixtures("skip_xp_backends"),]
+
+IS_WINDOWS_AND_NP1 = os.name == 'nt' and np.__version__ < '2'
```
There is no issue. It's just there is a numpy-only test of an internal routine (and everything internal is numpy-only) and numpy changed the default int dtype on Windows in NumPy 2.

lucascolley replied:
cool, so this is only needed because of the dtype checking introduced here? Makes sense 👍

ev-br commented on scipy/ndimage/tests/test_measurements.py:
```diff
@@ -31,8 +36,10 @@ def test_a(self):
-            assert_array_equal(counts, [2, 2])
-            assert_array_equal(sums, [1.0, 8.0])
+
+            dtype_arg = {'dtype': np.int64} if IS_WINDOWS_AND_NP1 else {}
+            xp_assert_equal(counts, np.asarray([2, 2], **dtype_arg))
+            xp_assert_equal(sums, np.asarray([1.0, 8.0]))
```
Yes. It's testing an internal helper from ndimage/_measurements.py, and it's numpy-only. I've added a comment to the test class docstring.

lucascolley replied:
sorry, what I meant was, can you mark the class with an `np_only` skip? I know that might seem too verbose as it means introducing unused `xp` args to the test methods, but it is useful to have all NumPy-only tests be discoverable with a single search.

ev-br reviewed (commented):
Thank you for the review Lucas! I believe I addressed all your comments.

ev-br commented:
Can we please flag what's essential to change and what can be left for a follow-up? At this point I'm no longer clear on what is what.

lucascolley commented:
> Can we please flag what's essential to change and what can be left for a follow-up? At this point I'm no longer clear on what is what.

Sure, I'll have to take another look through your round of changes first.

lucascolley reviewed (changes requested):
Thanks Evgeni, again this is a huge amount of work! I have resolved all addressed comments - it should be clear now what I expect to be changed before merge here. I've opened gh-21280 to track follow-ups.

ev-br commented:
OK, https://github.com/scipy/scipy/pull/21150/commits/5cfe3f09bc033a46155f15869718975f5ae73e62 addresses the two remaining comments. 

TBH, I don't necessarily agree that the `sumsq` change is that useful, on the grounds that it adds essentially dead code. But this is not essential enough for me to argue, and it's changed in the above commit. 

Let me know if you're happy with the result @lucascolley . The last thing is probably to squash the commits some --- 60+ commits is not useful here, so it's either a squash-merge (5kLOC) or separate into a pair of commits: one to add the delegation layer, the other to overhaul the tests (still 5kLOC). I'm leaning towards the second option, just so that the delegation stuff is more discoverable.

lucascolley reviewed (approved):
> Let me know if you're happy with the result @lucascolley . The last thing is probably to squash the commits some --- 60+ commits is not useful here, so it's either a squash-merge (5kLOC) or separate into a pair of commits: one to add the delegation layer, the other to overhaul the tests (still 5kLOC). I'm leaning towards the second option, just so that the delegation stuff is more discoverable.

LGTM! Squashing into a pair of commits sounds sensible to me 👍

ev-br committed:
ENH: ndimage: add the array API delegation layer

The internal ndimage implementation remains numpy-only, and is intact.

- On CuPy, delegate ndimage.func to cupyx.scipy.ndimage.func
- On other backends, convert inpit arguments to numpy arrays,
  run the numpy-only internal function, and convert the results back
- On JAX, delegate for map-coordinates and convert for the rest.

This way, each ndimage function gets its _dispatcher function, which
knows which arguments are array-like, and deduces the array namespace
based on inputs. This deduction stage needs to be function-specific.
The rest is regular, and is done in _support_alternative_backends.py

Constructing the public API goes through an indirection layer:

- _ndimage_api.py collect imports from private implementation modules;
- _support_alternative_backends.py handles decorates implementor functions;
- __init__.py import from _support_alternative_backends.py

ev-br committed:
MAINT: ndimage: add np.asarray calls, where missing

Since ndimage/_{implementation module} are all numpy-only, they
need to make sure their arguments are not cupy/torch/jax/etc arrays.

Previously only numpy arrays or python array-convertible structures
were possible, so several missing asarray calls were not a problem
and conversion was done ad hoc in numpy calls in the 'computational'
parts of the functions.

ev-br committed:
TST: ndimage: convert ndimage tests to the array API infrastructure

The diff is large but largely mechanical.

ev-br committed:
ENH: _lib: add backwards compatible look-alikes for assert_almost_equal

... and assert_array_almost equal.

This is to reduce the size of the diff when tests are converted to use
xp_assert_* infrastructure.

ev-br committed:
CI: ndimage: test alt backends on CI

Also silence several annoying but harmless warnings: cupy likes to warn that
jittify is experimental, mypy is acting up on the import structure etc;
there's no need to fail the tests suite on these.

ev-br commented:
Rebased. It's five commits in the end: delegation, changes to ndimage itself, _lib testing infra, ndimage/testing refactors, and CI + mypy, all in separate commits.

lucascolley commented:
Landed - thanks Evgeni!

lucascolley commented:
@mdhaber if you are still interested in looking at generalising the *delegation*[^1] infrastructure across here, `special` and beyond, that would be fantastic!

[^1]: we decided in the community meeting that avoiding the word "dispatching" to refer to this is probably best, now that work on dispatching systems seems to be picking back up.
